src/Components/Endpoints/src/Binding/FormDataReader.cs

```diff
@@ -9,57 +11,193 @@ namespace Microsoft.AspNetCore.Components.Endpoints.Binding;
 {
-    private readonly IReadOnlyDictionary<string, StringValues> _formCollection;
-    private string _prefix;
+    private readonly IReadOnlyDictionary<FormKey, StringValues> _readOnlyMemoryKeys;
+    private readonly Memory<char> _prefixBuffer;
+    private Memory<char> _currentPrefixBuffer;
 
-    public FormDataReader(IReadOnlyDictionary<string, StringValues> formCollection, CultureInfo culture)
+    // As an implementation detail, reuse FormKey for the values.
+    // It's just a thin wrapper over ReadOnlyMemory<char> that caches
+    // the computed hash code.
+    private IReadOnlyDictionary<FormKey, HashSet<FormKey>>? _formDictionaryKeysByPrefix;
+
+    public FormDataReader(IReadOnlyDictionary<FormKey, StringValues> formCollection, CultureInfo culture, Memory<char> buffer)
     {
-        _formCollection = formCollection;
-        _prefix = "";
+        _readOnlyMemoryKeys = formCollection;
         Culture = culture;
+        _prefixBuffer = buffer;
     }
 
+    internal ReadOnlyMemory<char> CurrentPrefix => _currentPrefixBuffer;
+
     public IFormatProvider Culture { get; internal set; }
 
-    internal IEnumerable<string> GetKeys()
+    public Action<string, FormattableString, string?>? ErrorHandler { get; set; }
+
+    public void AddMappingError(FormattableString errorMessage, string? attemptedValue)
+    {
+        ArgumentNullException.ThrowIfNull(errorMessage);
+
+        if (ErrorHandler == null)
+        {
+            throw new FormDataMappingException(new FormDataMappingError(_currentPrefixBuffer.ToString(), errorMessage, attemptedValue));
+        }
+
+        ErrorHandler.Invoke(_currentPrefixBuffer.ToString(), errorMessage, attemptedValue);
+    }
+
+    internal FormKeyCollection GetKeys()
+    {
+        if (_formDictionaryKeysByPrefix == null)
+        {
+            _formDictionaryKeysByPrefix = ProcessFormKeys();
+        }
+
+        if (_formDictionaryKeysByPrefix.TryGetValue(new FormKey(_currentPrefixBuffer), out var foundKeys))
+        {
+            return new FormKeyCollection(foundKeys);
+        }
+
+        return FormKeyCollection.Empty;
+    }
+
+    // Internal for testing purposes
+    internal IReadOnlyDictionary<FormKey, HashSet<FormKey>> ProcessFormKeys()
     {
-        return _formCollection.Keys;
+        var keys = _readOnlyMemoryKeys.Keys;
+        var result = new Dictionary<FormKey, HashSet<FormKey>>();
+        // We need to iterate over all the keys in the dictionary and process each key to split it into segments where
+        // the prefixes are string separated by . and the keys are enclosed in []. For example if the key is
+        // Customer.Orders[<<OrderId>>]BillingInfo.FirstName, then we need to split it into Customer.Orders,
+        // [<<OrderId>>] and BillingInfo.FirstName. We then, need to group all the keys by the prefix. So, for the
+        // above example, we will have an entry for the prefix Customer.Orders that will include [<<OrderId>>] as the
+        // key.
```

Comment:
Am I right to think that, later once the model binding system is able to supply the messages in the original `FieldIdentifier` format (object+propertyname pairs) we won't need to receive or process any data in this format? Or will this still be used for some other reason?

Reply:
Based on conversations, I think this can be made work with some extra processing on the binder but will tackle it later as it's not trivial. I'll file an issue.

Reply:
This is done now.